### PR TITLE
CP-46851: add parameter to skip device types on get_export_metadata 

### PR DESCRIPTION
**Pull request by psafont**
This allows orchestrators to copy devices like VBDs from the VM, export the
metadata without the disk, import the VM to any pool, and finally add the
saved VBDs to the new VM before booting it.

Also include cleanups in the cli server, it's full of ad-hoc methods.

This has passed manual testing, the corresponding FQP test by QA, and tested by CVAD people as well

### CODE REVIEW

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 1]
```diff
@@ -5860,14 +5854,21 @@ let blob_create printer rpc session_id params =
 
 let export_common fd _printer rpc session_id params filename num ?task_uuid
     compression preserve_power_state vm =
-  let vm_metadata_only : bool = get_bool_param params "metadata" in
-  let export_snapshots : bool =
-    if List.mem_assoc "include-snapshots" params then
-      bool_of_string "include-snapshots" (List.assoc "include-snapshots" params)
+  let vm_metadata_only = get_bool_param params "metadata" in
+  let export_snapshots = get_bool_param params "include-snapshots" in
+  let uri =
```
Comment (lindig) [inline ocaml/xapi-cli-server/cli_operations.ml:5859]: Just an idea:
```
let uri, extra_args = match  vm_metadata_only with
| true -> Constants.export_metadata_uri,  Printf.sprintf "&excluded_device_types=%s"
        (get_param params ~default:"" "excluded-device-types")
| false ->  Constants.export_uri, ""
```

File: ocaml/xapi/import.ml [hunk 2]
```diff
@@ -2158,11 +2158,18 @@ let complete_import ~__context vmrefs =
         Xapi_vm_lifecycle.update_allowed_operations ~__context ~self:vm
       )
       vmrefs ;
-    (* We only keep VMs which are not snapshot *)
+    (* When only snapshots have been imported, return all of them.
```
Comment (lindig) [inline ocaml/xapi/import.ml:2161]: I don't have the context of this - but is this a behavior that customer expect?

File: ocaml/xapi/importexport.ml [hunk 3]
```diff
@@ -469,6 +469,39 @@ module Format = struct
...
+    | "vif" ->
+        VIF
+    | "vbd" ->
+        VBD
+    | "vgpu" ->
+        VGPU
+    | "vtpm" ->
+        VTPM
+    | other ->
+        let msg =
```
Comment (lindig) [inline ocaml/xapi/importexport.ml:498]:
```
let fail fmt = Printf.kprintf failwith fmt in
fail "%s: Type '%s' not one of [%s]" __FUNCTION__ other
            (String.concat "; " (List.map to_string all))
```

**Review by lindig — APPROVED** [review]

File: ocaml/xapi/import.ml (hunk 2 above)
Comment (lindig) [reply ocaml/xapi/import.ml:2161]: I was more confused by the difference between when only snapshots are imported and the case were a mix of snapshots and VMs are imported - as I read the comment, the behaviour is different.

Comment (psafont) [reply ocaml/xapi/import.ml:2161]: Before this PR the export of metadata-only was blocked, so I don't think this change of behaviour affects any users :)

File: ocaml/xapi/export.ml [hunk 4]
```diff
@@ -118,15 +120,16 @@ let rec update_table ~__context ~include_snapshots ~preserve_power_state
            )
        ) ;
     (* add VTPMs that belong to this VM *)
-    vm.API.vM_VTPMs
-    |> List.iter (fun ref -> if Db.is_valid_ref __context ref then add ref) ;
+    if not (List.mem Devicetype.VTPM excluded_devices) then
```
Comment (Vincent-lau) [inline ocaml/xapi/export.ml:123]: Do we allow users to exclude vtpms as well (and then add them back later on)?

**Review by Vincent-lau — APPROVED** [review]

File: ocaml/xapi/export.ml (hunk 4 above)
Comment (psafont) [reply ocaml/xapi/export.ml:123]: Yes, this code is doing the exclusion. For adding it back to the VM, I don't think there's an easy way to export the TPM contents separately

File: ocaml/xapi-cli-server/cli_operations.ml [hunk 5]
```diff
@@ -3310,11 +3311,11 @@ let do_host_op rpc session_id op params ?(multiple = true) ignore_params =
...
-  match List.length srs with
-  | 0 ->
+  match srs with
+  | [] ->
       failwith "No matching hosts found"
-  | 1 ->
-      [op (List.hd srs)]
+  | [sr] ->
+      [op sr]
   | _ ->
```
[the remark this answers is not in the document]
Comment (psafont) [reply ocaml/xapi-cli-server/cli_operations.ml:3319]: This file is in need of a lot of care to use sensible coding standards. I think removing the use of List.hd is enough for now. Maybe as an OIL task...

**Pull request merged**